### PR TITLE
Add database sync auth token

### PR DESCRIPTION
**Pull request by agustarnason**

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncTokenService.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            .WithAuthority(new Uri($"https://login.microsoftonline.com/{optionValues.TenantId}"))
+            .Build();
+    }
+
+    public async Task<string> GetAccessTokenAsync()
+    {
+        try
+        {
+            var optionValues = _options.CurrentValue;
+            var result = await _app.AcquireTokenForClient([optionValues.Scope]).ExecuteAsync();
```
Comment (TordJoranger) [inline src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncTokenService.cs:26]: I would expect us to use AquireTokenAsync here

**Review by TordJoranger — APPROVED** [review]

**Review by janid1967 — CHANGES_REQUESTED** [review]
merge main in, SyncToPCS4Options is moved to Equinor.ProCoSys.Completion.Domain

**Comment by agustarnason** [comment]
> merge main in, SyncToPCS4Options is moved to Equinor.ProCoSys.Completion.Domain

Are these change on main, or are they not yet merged?
I can't seem to find SyncToPCS4Options in Domain
Or did you mean for SyncToPCS4Options to be merged into ApplicationOptions?

**Review by janid1967 — APPROVED** [review]

**Comment by janid1967** [comment]
Please update https://statoildeveloper.visualstudio.com/Statoil.NextGenProCoSys/_wiki/wikis/Statoil.NextGenProCoSys.wiki/1910/Completion-API with new SyncToPCS4Options.Scope

File: src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncBearerTokenHandler.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+    public SyncBearerTokenHandler(ISyncTokenService syncTokenService)
+    {
+        _syncTokenService = syncTokenService;
+    }
+
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        var token = await _syncTokenService.AquireTokenAsync();
+        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
```
Comment (DownGoat) [inline src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncBearerTokenHandler.cs:17]: We should cache this token

**Review by DownGoat — APPROVED** [review]

File: src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncBearerTokenHandler.cs (hunk 2 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncBearerTokenHandler.cs:17]: The same improvement should be done when we chat with main api too... that code is in Common repo ... if doing same improvement in Common for main api chatting, consider to make code general in Common so we don't need SyncBearerTokenHandler.cs at all (PS: Preservation chat with both Main api + Library api, and use same code to get token and create a http client)

File: src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncBearerTokenHandler.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        _cacheManager = cacheManager;
+    }
+
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        var token = await _cacheManager.GetOrCreate(
+            "SYNC_TOKEN",
+            async () =>
+            {
+                var checkList = await _syncTokenService.AquireTokenAsync();
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncBearerTokenHandler.cs:24]: rename to token

**Review by janid1967 — CHANGES_REQUESTED** [review]

File: src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncBearerTokenHandler.cs (hunk 2 above)
Comment (agustarnason) [reply src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncBearerTokenHandler.cs:29]: AquireTokenAsync itself calls AquireTokenAsync via IAuthenticatorOptions, and by default this function caches the token
The token is cached for as long as the token is valid, i.e. until expiration

Comment (agustarnason) [reply src/Equinor.ProCoSys.Completion.DbSyncToPCS4/Service/SyncBearerTokenHandler.cs:17]: If custom caching for this token is required, then it should probably be implemented in a different PR

**Review by janid1967 — APPROVED** [review]

**Pull request merged**